src/main/scala/org/ergoplatform/network/ErgoNodeViewSynchronizer.scala

```diff
@@ -899,30 +984,39 @@ class ErgoNodeViewSynchronizer(networkControllerRef: ActorRef,
       broadcastInvForNewModifier(mod)
       if (mod.isInstanceOf[ErgoFullBlock]) {
         clearDeclined()
+        clearTransactionsInfo()
+        processFirstTxProcessingCacheRecord() // resume cache processing
       }
 
-    case SuccessfulTransaction(tx) =>
+    case st@SuccessfulTransaction(utx) =>
+      val tx = utx.transaction
       deliveryTracker.setHeld(tx.id, Transaction.ModifierTypeId)
+      processMempoolResult(st)
       broadcastModifierInv(tx)
 
-    case DeclinedTransaction(id: ModifierId) =>
-      declined.put(id, System.currentTimeMillis())
-
-    case FailedTransaction(id, error, immediateFailure) =>
-      if (immediateFailure) {
-        // penalize sender only in case transaction was invalidated at first validation.
-        deliveryTracker.setInvalid(id, Transaction.ModifierTypeId).foreach { peer =>
-          error match {
-            case TooHighCostError(_) =>
-              log.info(s"Penalize spamming peer $peer for too costly transaction $id")
-              penalizeSpammingPeer(peer)
-            case _ =>
-              log.info(s"Penalize peer $peer for too costly transaction $id (reason: $error)")
-              penalizeMisbehavingPeer(peer)
-          }
+    case dt@DeclinedTransaction(utx: UnconfirmedTransaction) =>
+      declined.put(utx.id, System.currentTimeMillis())
+      processMempoolResult(dt)
+
+    case ft@FailedTransaction(utx, error) =>
+      val id = utx.id
+      processMempoolResult(ft)
+
+      utx.source.foreach { peer =>
+        // no need to call deliveryTracker.setInvalid, as mempool will consider invalidated tx in contains()
+        error match {
+          case TooHighCostError(_) =>
+            log.info(s"Penalize spamming peer $peer for too costly transaction $id")
+            penalizeSpammingPeer(peer)
+          case _ =>
+            log.info(s"Penalize peer $peer for too costly transaction $id (reason: $error)")
+            penalizeMisbehavingPeer(peer)
         }
       }
 
+    case FailedOnRecheckTransaction(_, _) =>
```

Comment:
Shouldn't processFirstTxProcessingCacheRecord be called here?

Reply:
I guess no, as re-checking is a separate process with own limits, so not related to mempool backpressure